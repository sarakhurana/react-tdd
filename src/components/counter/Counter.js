import React, { useState } from "react";

const Counter = () => {
  const [counter, setCounter] = useState(0);

  const handleIncrement = () => {
    setTimeout(() => setCounter(counter + 1), 5000);
  };
  const handleDecrement = () => {
    setCounter(counter - 1);
  };
  return (
    <>
      <h1 data-testid="data-test-counter-value">{counter}</h1>
      <button data-testid="data-test-increment-btn" onClick={handleIncrement}>
        +
      </button>
      <button data-testid="data-test-decrement-btn" onClick={handleDecrement}>
        -
      </button>
    </>
  );
};
export default Counter;
