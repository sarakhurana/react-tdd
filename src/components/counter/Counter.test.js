import { render, act, fireEvent, waitFor } from "@testing-library/react";
import Counter from "./Counter";

describe("Basic rendering of Counter", () => {
  test("should render", () => {
    const { asFragment } = render(<Counter />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("should show counter text", () => {
    const { getByTestId, getByText } = render(<Counter />);

    const counterValue = getByTestId("data-test-counter-value");

    expect(counterValue.textContent).toBe("0");
    expect(getByText("0")).toBeInTheDocument();
  });

  test("should render increment button", () => {
    const { getByTestId } = render(<Counter />);

    const incrementBtn = getByTestId("data-test-increment-btn");

    expect(incrementBtn).toBeDefined();
  });

  test("should render decrement button as disabled", () => {
    const { getByTestId } = render(<Counter />);

    const decrementBtn = getByTestId("data-test-decrement-btn");

    expect(decrementBtn).toBeDefined();
  });
});

describe("Counter functionality", () => {
  test("increment button should increase counter value by 1", async() => {
    //jest.useFakeTimers();
    const { getByTestId, getByText } = render(<Counter />);

    const incrementBtn = getByTestId("data-test-increment-btn");
     act(()=>{fireEvent.click(incrementBtn)})
     //jest.runAllTimers();

     const counter = await waitFor(()=>getByText("1"));

     expect(counter).toHaveTextContent("1");
  });

  test("decrement button should decrease counter value by 1", () => {
    const { getByTestId } = render(<Counter />);

    const decrementBtn = getByTestId("data-test-decrement-btn");
    act(()=>{fireEvent.click(decrementBtn)})

    expect(getByTestId("data-test-counter-value").textContent).toBe("-1");
  });
});
