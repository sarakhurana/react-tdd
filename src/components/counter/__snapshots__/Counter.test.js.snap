// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic rendering of Counter should render 1`] = `
<DocumentFragment>
  <h1
    data-testid="data-test-counter-value"
  >
    0
  </h1>
  <button
    data-testid="data-test-increment-btn"
  >
    +
  </button>
  <button
    data-testid="data-test-decrement-btn"
  >
    -
  </button>
</DocumentFragment>
`;
